### PR TITLE
Simplify consumption side of the new async navigation helpers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/59802.

Continues work on async navigation, removing old helpers.

THe core idea here is that we'd like the INavigableLocation to potentially be synchronous to navigate, indicating that all expensive BG work has already been done, and the only remaining work is a quite opening of the UI to take the user to the actual location.  As such, all consumers of it must get the location asynchronously, then switch to the UI as the very last step, then do the actual navigation.

Note; tehre is only one final portion of the code preventing the actual switch rom async to sync here.  That's the reentrancy in MAS where we use async to first generate the MAS file, then we try to open it, reentering to call back into navigation to find the doc post open.  Once we clean that up, the final swithc can happen.

### CODE REVIEW

File: src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs [hunk 1]
```diff
@@ -58,36 +58,25 @@ private class NavigableSymbol : INavigableSymbol
...
 
-                    var cancellationToken = context.UserCancellationToken;
-                    var location = await _presenter.GetStreamingLocationAsync(
-                        _threadingContext,
-                        _document.Project.Solution.Workspace,
-                        _definitions[0].NameDisplayParts.GetFullText(),
-                        _definitions,
-                        cancellationToken).ConfigureAwait(false);
-                    if (location != null)
-                        await location.NavigateToAsync(new NavigationOptions(PreferProvisionalTab: true, ActivateTab: true), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/NavigableSymbols/NavigableSymbolService.NavigableSymbol.cs:83]: note: i don't pass these options down as all callers use this nav value for the TryPresentLocationsAsync calls.  so it's just specified inside that instead.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski this is ready for review.

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs [hunk 2]
```diff
@@ -301,7 +300,7 @@ internal class CodeActionEditHandlerService : ForegroundThreadAffinitizedObject,
             {
                 var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
                 await navigationService.TryNavigateToPositionAsync(
-                    workspace, navigationOperation.DocumentId, navigationOperation.Position, cancellationToken).ConfigureAwait(false);
+                    this.ThreadingContext, workspace, navigationOperation.DocumentId, navigationOperation.Position, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:303]: teh general concept here is that all the helpers which actually do perform a navigation take a ThreadingCOntext.  That way they *could*, if necessary, switch to a UI thread.  This also means taht the helpers are only available to consumer who ahve a threading context and thus live in an 'editor' world where the concepts of navigating actually make sense.

File: src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptDocumentNavigationServiceWrapper.cs [hunk 3]
```diff
@@ -35,13 +35,25 @@ public bool TryNavigateToPosition(Workspace workspace, DocumentId documentId, in
         public bool TryNavigateToPosition(Workspace workspace, DocumentId documentId, int position, int virtualSpace, OptionSet? options, CancellationToken cancellationToken)
         {
             var obj = _underlyingObject;
-            return _threadingProvider.Service.Run(() => obj.TryNavigateToPositionAsync(workspace, documentId, position, virtualSpace, NavigationOptions.Default, cancellationToken));
+            return _threadingProvider.Service.Run(async () =>
+            {
+                var location = await obj.GetLocationForPositionAsync(
+                    workspace, documentId, position, virtualSpace, cancellationToken).ConfigureAwait(false);
+                return location != null &&
+                    await location.NavigateToAsync(NavigationOptions.Default, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptDocumentNavigationServiceWrapper.cs:43]: this is a genuine weird case.  the TS code lives at the feature layer, but has hte concept of navigation.  if we changes NavigateTo to be sync and require the UI thread, we'd have to introduce some sort of hack (likely through the IWorkspaceThreadingService) to somehow allow it ot move over.  I don't have a good idea of what we can do here that isn't hack.  Thoughts @jasonmalinowski ?

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemDisplay.cs [hunk 4]
```diff
@@ -117,13 +117,17 @@ public void NavigateTo()
...
             // TODO: Get the platform to use and pass us an operation context, or create one
             // ourselves.
-            _threadingContext.JoinableTaskFactory.Run(() => navigationService.TryNavigateToSpanAsync(
-                workspace,
-                document.Id,
-                _searchResult.NavigableItem.SourceSpan,
-                NavigationOptions.Default,
-                allowInvalidSpan: _searchResult.NavigableItem.IsStale,
-                CancellationToken.None));
+            _threadingContext.JoinableTaskFactory.Run(async () =>
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemDisplay.cs:120]: Hmm, why the await here?

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:303]: Ah, interesting way to split that up. I think I like it!

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs [hunk 5]
```diff
@@ -342,11 +341,9 @@ internal class CodeActionEditHandlerService : ForegroundThreadAffinitizedObject,
                         {
                             var editorWorkspace = workspace;
                             var navigationService = editorWorkspace.Services.GetRequiredService<IDocumentNavigationService>();
-                            var location = await navigationService.GetLocationForSpanAsync(
-                                editorWorkspace, documentId, resolvedRenameToken.Span, cancellationToken).ConfigureAwait(false);
 
-                            if (location != null &&
-                                await location.NavigateToAsync(NavigationOptions.Default, cancellationToken).ConfigureAwait(false))
+                            if (await navigationService.TryNavigateToSpanAsync(
+                                    this.ThreadingContext, editorWorkspace, documentId, resolvedRenameToken.Span, cancellationToken).ConfigureAwait(false))
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:346]: Indenting intentional this way?

File: src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptDocumentNavigationServiceWrapper.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/Features/Core/Portable/ExternalAccess/VSTypeScript/Api/VSTypeScriptDocumentNavigationServiceWrapper.cs:43]: Is there a reason the TypeScript stuff isn't at the editor features layer?

File: src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItem.cs [hunk 6]
```diff
@@ -51,29 +51,37 @@ public override object GetBrowseObject()
...
-                        // TODO: Use a threaded-wait-dialog here so we can cancel navigation.
-                        didNavigate |= item._threadingContext.JoinableTaskFactory.Run(() =>
-                            documentNavigationService.TryNavigateToPositionAsync(item.Workspace, item.DocumentId, position: 0, CancellationToken.None));
+                        var documentNavigationService = item.Workspace.Services.GetService<IDocumentNavigationService>();
+                        if (documentNavigationService != null)
+                        {
+                            // TODO: we're navigating back to the top of the file, do we have a way to just bring it to the focus and that's it?
+                            // TODO: Use a threaded-wait-dialog here so we can cancel navigation.
+                            didNavigate |= await documentNavigationService.TryNavigateToPositionAsync(
+                                _threadingContext, item.Workspace, item.DocumentId, position: 0, CancellationToken.None).ConfigureAwait(false);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItem.cs:79]: Before you were using item._threadingContext, did you not want to keep using that?

**Review by jasonmalinowski — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Going to merge this in.  But address all the feedback in a quick followup as i think it's all relevant.

File: src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs [hunk 7]
```diff
@@ -53,6 +53,20 @@ internal interface IStreamingFindUsagesPresenter
...
+        {
+            var location = await presenter.GetStreamingLocationAsync(
+                threadingContext, workspace, title, items, cancellationToken).ConfigureAwait(false);
+            return await location.NavigateToAsync(
+                threadingContext, new NavigationOptions(PreferProvisionalTab: true, ActivateTab: true), cancellationToken).ConfigureAwait(false);
+        }
+
         /// <summary>
         /// If there's only a single item, navigates to it.  Otherwise, presents all the
         /// items to the user.
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs:72]: will do!

File: src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs [hunk 8]
```diff
@@ -53,6 +53,20 @@ internal interface IStreamingFindUsagesPresenter
 
     internal static class IStreamingFindUsagesPresenterExtensions
     {
+        public static async Task<bool> TryPresentLocationsAsync(
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Host/IStreamingFindReferencesPresenter.cs:56]: sounds good.  i like that name :)

File: src/EditorFeatures/Core/Navigation/IDocumentNavigationServiceExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Navigation
+{
+    internal static class INavigableLocationExtensions
+    {
+        public static async Task<bool> NavigateToAsync(
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Navigation/IDocumentNavigationServiceExtensions.cs:14]: sure.  that seems reasonable.

File: src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemDisplay.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core.Wpf/NavigateTo/NavigateToItemDisplay.cs:120]: godo catch.  no reason will fix.

File: src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs:346]: yup.  i dont' like my arguments aligning with the parent, they are better indented to visually tell tehy're children, not siblings :)

File: src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItem.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItem.cs:79]: i likely missed that.  it's probably fine either way, but i don't mind switching back.

**Comment by CyrusNajmabadi** [comment]
> Is there a reason the TypeScript stuff isn't at the editor features layer?

:'(

**Pull request merged**